### PR TITLE
Testcase for Nimble-python-sdk developed using pytest framework

### PR DESCRIPTION
**Pull request by ar-india**
The testcase has been implemented using pytest module.
-some of the review comments like indentation of code, following snake_case for function and variables name have been followed.
-have used flake as lint tool. Thanks to cris for letting me know. it's really nice tool

Please review the file.

### CODE REVIEW

File: tests/config.ini [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#Do NOT change the section name
+
+# please provide the Nimble Array credentials against which the test case will be run. Note that the password is plain text and hence when the same is released to public github it will be removed
+[NimbleArrayCredentials] #section name
+hostname: 10.18.161.229
+#hostname: 10.18.171.96
+username: admin
+password: admin
```
Comment (ar-india) [inline tests/config.ini:8]: password will be blacked out in final push to public repo

File: tests/nimbleclientbase.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+from nimbleclient.v1 import client
+import datetime
+import time
+import configparser
+import os
+import threading
+
+
+# the below variable "SKIPTEST" is to be used if a user wants to just run one
```
Comment (rgcostea) [inline tests/nimbleclientbase.py:15]: nit: The

File: tests/nimbleclientbase.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+# particular function.They should set the value of this to 1  on command
+# prompt and then change the value of SKIPTEST to false for the function they
+# wish to debug.If they want to skip the entire tests in this testcase, then
```
Comment (rgcostea) [inline tests/nimbleclientbase.py:18]: to debug.  If they

File: tests/nimbleclientbase.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+from nimbleclient.v1 import client
+import datetime
+import time
+import configparser
+import os
+import threading
+
+
+# the below variable "SKIPTEST" is to be used if a user wants to just run one
+# particular function.They should set the value of this to 1  on command
```
Comment (rgcostea) [inline tests/nimbleclientbase.py:16]: function.  They

and extra space after "to 1"

File: tests/test_volume.py [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+    def after_running_all_testcase():
+        log("**** Completed Tests for Volume TestCase *****")
+    request.addfinalizer(after_running_all_testcase)
+
+
+def cleanup_old_volumes():
+
+    # due to bulk_move operation most of the times when abort is issued.
+    # the array ignores the request to delete the volume as the operation
+    # is already in progress and hence the volumes are left over.hence,
```
Comment (rgcostea) [inline tests/test_volume.py:35]: left over.  Hence,

File: tests/test_volume.py [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+    cleanup_old_volumes()
+
+    def after_running_all_testcase():
+        log("**** Completed Tests for Volume TestCase *****")
+    request.addfinalizer(after_running_all_testcase)
+
+
+def cleanup_old_volumes():
+
+    # due to bulk_move operation most of the times when abort is issued.
```
Comment (rgcostea) [inline tests/test_volume.py:33]: Due

File: tests/test_volume.py [hunk 7]
```diff
@@ -0,0 +1,540 @@
...
+    # the array ignores the request to delete the volume as the operation
+    # is already in progress and hence the volumes are left over.hence,
+    # every time this test is run ,we will try to remove the old entries
```
Comment (rgcostea) [inline tests/test_volume.py:36]: run, we

File: tests/test_volume.py [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+
+@pytest.fixture(scope='function')
+def setup_teardown_for_each_test(before_running_all_testcase, request):
+    global vol_name1, vol_name2, vol_name3
+    # setup operations before yield is called
+    nimosclientbase.log_header(request.function.__name__)
+    yield setup_teardown_for_each_test
+    # teardown operations below
+    delete_volume()
+    # create a new volume names for the next testcase
```
Comment (rgcostea) [inline tests/test_volume.py:76]: create new volume names

File: tests/test_volume.py [hunk 9]
```diff
@@ -0,0 +1,540 @@
...
+        if"SM_too_large_page_size" in str(ex):
+            log("Failed as expected. Invaild pagesize given")
+        else:
+            log(ex)
+
+
+@pytest.mark.skipif(SKIPTEST is True,
+                    reason="skipped this test as SKIPTEST variable is true")
+def test_create_volume_readonly(setup_teardown_for_each_test):
+    # create a read only volume..the volume gets created as write only.
```
Comment (rgcostea) [inline tests/test_volume.py:194]: only volume.  The volume gets created as read/write

File: tests/test_volume.py [hunk 10]
```diff
@@ -0,0 +1,540 @@
...
+        if(vol1.attrs.get("name") == obj.attrs.get("name")):
+            assert obj.attrs.get("id") == vol1.attrs.get("id")
+            assert obj.attrs.get("size") == vol1.attrs.get("size")
+            # now check no other extra attributes were returned from server
+            # for this volume
+            # try fetching some attribute
+            assert obj.attrs.get("read_only") is None
+
+
+# below function will be implemented when sdk is ready for accepting filters
```
Comment (rgcostea) [inline tests/test_volume.py:236]: should we add a TODO?

File: tests/test_volume.py [hunk 11]
```diff
@@ -0,0 +1,540 @@
...
+        resp = nimosclientbase.get_nimos_client().volumes.get(
+            startRow=1, endRow=1)
+        assert resp is None
+    except exceptions.NimOSAPIError as ex:
+        log(ex)
+
+@pytest.mark.skipif(SKIPTEST is True,
+                    reason="skipped this test as SKIPTEST variable is true")
+def test_update_volume_size_attribute(setup_teardown_for_each_test):
+    # first atleast create few volume
```
Comment (rgcostea) [inline tests/test_volume.py:322]: rewrite this comment

File: tests/test_volume.py [hunk 12]
```diff
@@ -0,0 +1,540 @@
...
+    except exceptions.NimOSAPIError as ex:
+        if 'SM_invalid_keyvalue' in str(ex):
+            log("Failed as expected")
+
+
+@pytest.mark.skipif(SKIPTEST is True,
+                    reason="skipped this test as SKIPTEST variable is true")
+def test_update_resize_volume_force(setup_teardown_for_each_test):
+    # shrinking an online volume is an operation which will fail
+    # unless force=true is specified.this tests the
```
Comment (rgcostea) [inline tests/test_volume.py:378]: specified.  This

File: tests/test_volume.py [hunk 13]
```diff
@@ -0,0 +1,540 @@
...
+    # forceupdateobject functionality.
```
Comment (rgcostea) [inline tests/test_volume.py:379]: force update object

File: tests/test_volume.py [hunk 14]
```diff
@@ -0,0 +1,540 @@
...
+
+    # first atleast create few volume
```
Comment (rgcostea) [inline tests/test_volume.py:381]: rewrite this comment

File: tests/test_volume.py [hunk 15]
```diff
@@ -0,0 +1,540 @@
...
+            "use force option to proceed." in str(ex)
+    # try with force option
+    resp = nimosclientbase.get_nimos_client().volumes.update(
+        id=resp.attrs.get("id"), size=5, force=True)
+    assert resp.attrs.get("size") == 5
+
+
+@pytest.mark.skipif(SKIPTEST is True,
+                    reason="skipped this test as SKIPTEST variable is true")
+def test_create_clone_volume(setup_teardown_for_each_test):
```
Comment (rgcostea) [inline tests/test_volume.py:403]: just name this test_clone_volume

File: tests/test_volume.py [hunk 16]
```diff
@@ -0,0 +1,540 @@
...
+@pytest.mark.skipif(SKIPTEST is True,
+                    reason="skipped this test as SKIPTEST variable is true")
+def test_bulk_move_volume(setup_teardown_for_each_test):
+    try:
+        orig_vol_pool_name = ""
+        # first atleast create few volume
+        resp = create_volume(vol_name3, size=5)
+        orig_vol_pool_name = resp.attrs.get("pool_name")
+        # get the dest pool id.
+        ppol_resp = nimosclientbase.get_nimos_client().pools.list()
```
Comment (rgcostea) [inline tests/test_volume.py:467]: should this be pool_resp?

File: tests/test_volume.py [hunk 17]
```diff
@@ -0,0 +1,540 @@
...
+            log(ex)
+
+
+@pytest.mark.skipif(SKIPTEST is True,
+                    reason="skipped this test as SKIPTEST variable is true")
+def test_move_volume(setup_teardown_for_each_test):
+
+    try:
+
+        # first atleast create volume
```
Comment (rgcostea) [inline tests/test_volume.py:503]: rewrite comment

**Review by rgcostea — APPROVED** [review]
lgtm.  I've identified some minor nits.

File: tests/nimbleclientbase.py (hunk 2 above)
Comment (ar-india) [reply tests/nimbleclientbase.py:15]: have renamed to Below variable. removed "the" and renamed below to Below

File: tests/test_volume.py (hunk 9 above)
Comment (ar-india) [reply tests/test_volume.py:194]: But the volume gets created as write only.

File: tests/test_volume.py (hunk 10 above)
Comment (ar-india) [reply tests/test_volume.py:236]: added todo

File: tests/test_volume.py (hunk 11 above)
Comment (ar-india) [reply tests/test_volume.py:322]: fixed

File: tests/test_volume.py (hunk 12 above)
Comment (ar-india) [reply tests/test_volume.py:378]: fixed

File: tests/test_volume.py (hunk 16 above)
Comment (ar-india) [reply tests/test_volume.py:467]: yes

**Comment by ar-india** [comment]
Thanks George.. have made changes as per the comments

**Review by suneeth51 — APPROVED** [review]

**Pull request merged**